File: experiment-4/runs/language=go_model=claude-opus-4-8_tooling=none/rep1/internal/mcp/mcp_test.go
// Context
// -------
// Behaviour-Driven tests for the MCP JSON-RPC layer. They drive the server
// through an in-memory request stream (newline-delimited JSON-RPC) exactly as a
// real MCP client over stdio would, and assert the handshake, tool listing and
// tool invocation behaviour. The server is backed by the real datasets loaded
// once in TestMain.
package mcp

import (
	"bufio"
	"encoding/json"
	"os"
	"strings"
	"testing"

	"brazilian-soccer-mcp/internal/soccer"
)

var testGraph *soccer.Graph

func TestMain(m *testing.M) {
	dir, ok := soccer.FindDataDir("")
	if !ok {
		panic("could not locate data/kaggle for tests")
	}
	g, err := soccer.LoadGraph(dir)
	if err != nil {
		panic(err)
	}
	testGraph = g
	os.Exit(m.Run())
}

// run feeds the given JSON-RPC request lines through a fresh server and returns
// the decoded responses, keyed by request id.
func run(t *testing.T, requests ...string) map[float64]Response {
	t.Helper()
	in := strings.NewReader(strings.Join(requests, "\n") + "\n")
	var out strings.Builder
	srv := NewServer(testGraph)
	if err := srv.Serve(in, &out); err != nil {
		t.Fatalf("Serve error: %v", err)
	}

	responses := map[float64]Response{}
	scanner := bufio.NewScanner(strings.NewReader(out.String()))
	scanner.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)
	for scanner.Scan() {
		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}
		var r Response
		if err := json.Unmarshal(line, &r); err != nil {
			t.Fatalf("bad response line %q: %v", line, err)
		}
		var id float64
		if len(r.ID) > 0 {
			_ = json.Unmarshal(r.ID, &id)
		}
		responses[id] = r
	}
	return responses
}

// callText extracts the text content of a tools/call response.
func callText(t *testing.T, r Response) string {
	t.Helper()
	if r.Error != nil {
		t.Fatalf("unexpected RPC error: %+v", r.Error)
	}
	b, _ := json.Marshal(r.Result)
	var res CallToolResult
	if err := json.Unmarshal(b, &res); err != nil {
		t.Fatalf("decode CallToolResult: %v", err)
	}
	if len(res.Content) == 0 {
		t.Fatal("empty tool content")
	}
	return res.Content[0].Text
}

func TestInitializeHandshake(t *testing.T) {
	// Given a client, When it sends initialize, Then the server returns its
	// protocol version, tool capability and identity.
	resps := run(t, `{"jsonrpc":"2.0","id":1,"method":"initialize","params":{}}`)
	r := resps[1]
	if r.Error != nil {
		t.Fatalf("initialize errored: %+v", r.Error)
	}
	b, _ := json.Marshal(r.Result)
	var init InitializeResult
	if err := json.Unmarshal(b, &init); err != nil {
		t.Fatal(err)
	}
	if init.ProtocolVersion != ProtocolVersion {
		t.Errorf("protocol version = %q, want %q", init.ProtocolVersion, ProtocolVersion)
	}
	if init.Capabilities.Tools == nil {
		t.Error("expected tools capability")
	}
	if init.ServerInfo.Name == "" {
		t.Error("expected server name")
	}
}

func TestNotificationProducesNoResponse(t *testing.T) {
	// Given an initialized notification (no id), When sent, Then no response is
	// emitted for it.
	resps := run(t,
		`{"jsonrpc":"2.0","method":"notifications/initialized"}`,
		`{"jsonrpc":"2.0","id":7,"method":"ping"}`,
	)
	if _, ok := resps[0]; ok {
		t.Error("notification should not produce a response")
	}
	if _, ok := resps[7]; !ok {
		t.Error("expected response to ping")
	}
}

func TestToolsList(t *testing.T) {
	// Given the server, When tools/list is called, Then the expected tools are
	// advertised, each with a name, description and input schema.
	resps := run(t, `{"jsonrpc":"2.0","id":2,"method":"tools/list"}`)
	b, _ := json.Marshal(resps[2].Result)
	var list ToolsListResult
	if err := json.Unmarshal(b, &list); err != nil {
		t.Fatal(err)
	}
	want := map[string]bool{
		"search_matches": true, "head_to_head": true, "team_stats": true,
		"standings": true, "search_players": true, "competition_stats": true,
		"list_metadata": true,
	}
	got := map[string]bool{}
	for _, tool := range list.Tools {
		got[tool.Name] = true
		if tool.Description == "" {
			t.Errorf("tool %s missing description", tool.Name)
		}
		if len(tool.InputSchema) == 0 {
			t.Errorf("tool %s missing input schema", tool.Name)
		}
	}
	for name := range want {
		if !got[name] {
			t.Errorf("missing expected tool %q", name)
		}
	}
}

func TestToolCallStandings(t *testing.T) {
	// Given the server, When standings for 2019 is requested via tools/call,
	// Then the formatted table names Flamengo as champion.
	resps := run(t, `{"jsonrpc":"2.0","id":3,"method":"tools/call","params":{"name":"standings","arguments":{"season":2019,"limit":3}}}`)
	text := callText(t, resps[3])
	if !strings.Contains(text, "Flamengo") || !strings.Contains(text, "Champion") {
		t.Errorf("standings text missing champion:\n%s", text)
	}
}

func TestToolCallHeadToHead(t *testing.T) {
	resps := run(t, `{"jsonrpc":"2.0","id":4,"method":"tools/call","params":{"name":"head_to_head","arguments":{"team_a":"Flamengo","team_b":"Fluminense"}}}`)
	text := callText(t, resps[4])
	if !strings.Contains(text, "Head-to-head") {
		t.Errorf("unexpected h2h text:\n%s", text)
	}
}

func TestToolCallSearchPlayers(t *testing.T) {
	resps := run(t, `{"jsonrpc":"2.0","id":5,"method":"tools/call","params":{"name":"search_players","arguments":{"name":"Neymar"}}}`)
	text := callText(t, resps[5])
	if !strings.Contains(text, "Neymar") {
		t.Errorf("expected Neymar in results:\n%s", text)
	}
}

func TestUnknownMethodReturnsError(t *testing.T) {
	// Given an unknown method, When called, Then a method-not-found error is
	// returned rather than crashing the server.
	resps := run(t, `{"jsonrpc":"2.0","id":9,"method":"does/not/exist"}`)
	if resps[9].Error == nil || resps[9].Error.Code != codeMethodNotFound {
		t.Errorf("expected method-not-found error, got %+v", resps[9])
	}
}

func TestUnknownToolReportsError(t *testing.T) {
	// Given an unknown tool name, When tools/call is invoked, Then the result
	// is flagged isError (an MCP-level, not protocol-level, failure).
	resps := run(t, `{"jsonrpc":"2.0","id":10,"method":"tools/call","params":{"name":"nope","arguments":{}}}`)
	b, _ := json.Marshal(resps[10].Result)
	var res CallToolResult
	_ = json.Unmarshal(b, &res)
	if !res.IsError {
		t.Errorf("expected isError for unknown tool, got %+v", res)
	}
}

File: experiment-14/runs/language=clojure_model=opus-4.8-fast_prompt=neutral/rep1/test/brsoccer/mcp_test.clj
;; =============================================================================
;; Tests for brsoccer.mcp -- the JSON-RPC / MCP protocol layer.
;; Exercises the pure handle-request dispatch and a full stdio round-trip.
;; =============================================================================
(ns brsoccer.mcp-test
  (:require [clojure.test :refer [deftest is testing]]
            [clojure.data.json :as json]
            [clojure.string :as str]
            [brsoccer.mcp :as mcp])
  (:import [java.io BufferedReader StringReader StringWriter]))

(deftest initialize-handshake
  (let [resp (mcp/handle-request {:id 1 :method "initialize" :params {}})]
    (is (= "2.0" (:jsonrpc resp)))
    (is (= 1 (:id resp)))
    (is (= mcp/protocol-version (get-in resp [:result :protocolVersion])))
    (is (= "brazilian-soccer-mcp" (get-in resp [:result :serverInfo :name])))))

(deftest tools-are-advertised
  (let [resp (mcp/handle-request {:id 2 :method "tools/list" :params {}})
        names (set (map :name (get-in resp [:result :tools])))]
    (is (contains? names "find_matches"))
    (is (contains? names "team_record"))
    (is (contains? names "search_players"))
    (is (contains? names "standings"))
    (is (contains? names "match_statistics"))
    (testing "every advertised tool has a description and input schema"
      (is (every? :description (get-in resp [:result :tools])))
      (is (every? :inputSchema (get-in resp [:result :tools]))))))

(deftest notifications-get-no-response
  (is (nil? (mcp/handle-request {:method "notifications/initialized"}))))

(deftest unknown-method-errors
  (let [resp (mcp/handle-request {:id 9 :method "does/not/exist" :params {}})]
    (is (= -32601 (get-in resp [:error :code])))))

(deftest tools-call-find-matches
  (let [resp (mcp/handle-request
               {:id 3 :method "tools/call"
                :params {:name "find_matches"
                         :arguments {:team "Flamengo" :opponent "Fluminense"}}})
        result (:result resp)]
    (is (false? (:isError result)))
    (is (str/includes? (get-in result [:content 0 :text]) "Head-to-head"))
    (is (some? (get-in result [:structuredContent :result])))))

(deftest tools-call-standings
  (let [resp (mcp/handle-request
               {:id 4 :method "tools/call"
                :params {:name "standings"
                         :arguments {:competition "Brasileirão Série A" :season 2019}}})
        text (get-in resp [:result :content 0 :text])]
    (is (str/includes? text "standings"))
    (is (str/includes? text "pts"))))

(deftest unknown-tool-is-flagged
  (let [resp (mcp/handle-request
               {:id 5 :method "tools/call"
                :params {:name "no_such_tool" :arguments {}}})]
    (is (true? (get-in resp [:result :isError])))))

(deftest full-stdio-round-trip
  (testing "newline-delimited JSON in -> JSON-RPC responses out; notifications produce nothing"
    (let [requests (str (json/write-str {:jsonrpc "2.0" :id 1 :method "initialize" :params {}}) "\n"
                        (json/write-str {:jsonrpc "2.0" :method "notifications/initialized"}) "\n"
                        (json/write-str {:jsonrpc "2.0" :id 2 :method "tools/list" :params {}}) "\n")
          in  (BufferedReader. (StringReader. requests))
          out (StringWriter.)]
      (mcp/serve! in out)
      (let [lines (->> (str/split-lines (str out)) (remove str/blank?))
            parsed (map #(json/read-str % :key-fn keyword) lines)]
        (testing "exactly two responses (the notification is not answered)"
          (is (= 2 (count parsed))))
        (is (= [1 2] (map :id parsed)))
        (is (= mcp/protocol-version (get-in (first parsed) [:result :protocolVersion])))
        (is (seq (get-in (second parsed) [:result :tools])))))))
